File: apps/web/lib/stage.js
// The layout contract's arithmetic, as an oracle — dev-docs/stage-contract.md.
//
// These functions produce the EXPECTED numbers: the safe content area a client offers, the
// reference box fit to its short axis, what the long axis does with the difference, and the fixed
// desktop window for a monitor. They are not the runtime. The stylesheet is the runtime
// (index.html's .stage / .screen rules and the container query), and the geometry tests measure
// that in a real engine against what these return. Keeping the arithmetic here, pure and tiny,
// is what lets the contract's fixture table be a tested claim rather than a hand-typed one:
// stage.test.mjs reads the table out of the document and checks every row against these.

/** The long side over the short side of the reference box: 4:3 landscape, 3:4 portrait. */
export const RATIO = 4 / 3;

/**
 * Fit the reference box into a client.
 *
 * @param {object} o
 * @param {number} o.width   viewport width, logical px
 * @param {number} o.height  viewport height, logical px
 * @param {{top?: number, right?: number, bottom?: number, left?: number}} [o.insets]
 *        the OS safe-area insets — status bar, Dynamic Island, home indicator
 * @param {{top?: number, bottom?: number}} [o.bars]
 *        the app's own bars — the top bar, and the bottom tab bar (portrait only; the caller
 *        decides, the chrome rule is not this function's claim)
 * @returns {{
 *   orientation: 'landscape' | 'portrait',
 *   safe: {w: number, h: number},   the stage: viewport less insets less bars
 *   ref: {w: number, h: number},    the reference box fit to the stage's short axis
 *   surplus: number,                long-axis room beyond the reference box — the sheet's to take
 *   paper: number,                  short-axis margin on EACH side when the long axis is too short
 * }}
 */
export function fitStage({ width, height, insets = {}, bars = {} }) {
  const { top = 0, right = 0, bottom = 0, left = 0 } = insets;
  const w = width - left - right;
  const h = height - top - bottom - (bars.top ?? 0) - (bars.bottom ?? 0);
  if (!(w > 0) || !(h > 0)) throw new RangeError(`no stage left: ${w}×${h} after insets and bars`);
  // A square counts as landscape: it is the orientation the desktop reference is designed in.
  const portrait = h > w;
  const [short, long] = portrait ? [w, h] : [h, w];
  const box = (s, l) => (portrait ? { w: s, h: l } : { w: l, h: s });
  const orientation = portrait ? 'portrait' : 'landscape';
  const refLong = short * RATIO;
  if (refLong <= long) {
    return { orientation, safe: { w, h }, ref: box(short, refLong), surplus: long - refLong, paper: 0 };
  }
  // The long axis cannot hold the reference: the box shrinks to it, and the short axis pads.
  const boxShort = long / RATIO;
  return { orientation, safe: { w, h }, ref: box(boxShort, long), surplus: 0, paper: (short - boxShort) / 2 };
}

/**
 * The desktop window's constants. Tuning values, not the contract: k is the share of the work
 * area the window asks for before the clamps; min/max bound the stage's long side; margin keeps
 * the window off the work area's edge when it is height-bound.
 */
export const DESKTOP = Object.freeze({
  landscape: Object.freeze({ k: 0.5, min: 840, max: 1600 }), // of work-area WIDTH; bounds stage width
  portrait: Object.freeze({ k: 0.9, min: 500, max: 1200 }), // of work-area HEIGHT; bounds stage height
  margin: 16,
});

const clamp = (v, lo, hi) => Math.min(Math.max(v, lo), hi);

/**
 * The fixed desktop window for a monitor, in logical px.
 *
 * @param {object} o
 * @param {number} o.workW  work-area width (monitor less menu bar / Dock / taskbar), logical px
 * @param {number} o.workH  work-area height, logical px
 * @param {number} o.bar    the app's title bar: 52 on macOS (overlay), 44 on Windows/Linux
 * @param {'landscape' | 'portrait'} o.orientation
 * @param {typeof DESKTOP} [o.constants]
 * @returns {{stage: {w: number, h: number}, window: {w: number, h: number}}}
 */
export function desktopWindow({ workW, workH, bar, orientation, constants = DESKTOP }) {
  const { margin } = constants;
  if (orientation === 'landscape') {
    const { k, min, max } = constants.landscape;
    const w = Math.min(clamp(k * workW, min, max), (workH - bar - margin) * RATIO);
    const h = w / RATIO;
    return { stage: { w, h }, window: { w, h: h + bar } };
  }
  const { k, min, max } = constants.portrait;
  const h = Math.min(clamp(k * workH, min, max), workH - bar - margin);
  const w = h / RATIO;
  return { stage: { w, h }, window: { w, h: h + bar } };
}

File: apps/desktop/src-tauri/src/stage.rs
//! The layout contract's desktop window — dev-docs/stage-contract.md.
//!
//! Pure arithmetic, no tauri: a monitor's work area in logical px, the app's title bar, and the
//! persisted orientation give a fixed window whose stage is exactly 4:3 (landscape) or 3:4
//! (portrait). The same formulas live in `apps/web/lib/stage.js` as the oracle the fixture tables
//! are checked against; the tests below hold this side to the document's desktop table.
//!
//! ```text
//! landscape:  w = min( clamp(minW, k·workW, maxW),  (workH − bar − margin) · 4/3 )   h = w · 3/4
//! portrait:   h = min( clamp(minH, k·workH, maxH),   workH − bar − margin )           w = h · 3/4
//! window    = (w, h + bar)
//! ```

/// The long side over the short side of the reference box.
pub const RATIO: f64 = 4.0 / 3.0;

/// Tuning values, not the contract: `k` is the share of the work area asked for before the
/// clamps, `min`/`max` bound the stage's long side.
#[derive(Clone, Copy, Debug)]
pub struct Constants {
    pub k: f64,
    pub min: f64,
    pub max: f64,
}

/// Of the work area's WIDTH; bounds the stage width.
pub const LANDSCAPE: Constants = Constants {
    k: 0.5,
    min: 840.0,
    max: 1600.0,
};
/// Of the work area's HEIGHT; bounds the stage height.
pub const PORTRAIT: Constants = Constants {
    k: 0.9,
    min: 500.0,
    max: 1200.0,
};
/// Keeps a height-bound window off the work area's edge.
pub const MARGIN: f64 = 16.0;

/// The app's own title bar, in logical px, drawn inside the window's content: 52 on macOS (the
/// overlay bar the traffic lights sit in), 44 on Windows and Linux (undecorated windows, the
/// caption buttons are the app's).
#[cfg(target_os = "macos")]
pub const BAR: f64 = 52.0;
#[cfg(not(target_os = "macos"))]
pub const BAR: f64 = 44.0;

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum Orientation {
    Landscape,
    Portrait,
}

impl Orientation {
    /// The persisted form. Anything else is landscape — the shape the app is designed in first.
    pub fn parse(s: &str) -> Self {
        if s.trim() == "portrait" {
            Self::Portrait
        } else {
            Self::Landscape
        }
    }

    pub fn as_str(self) -> &'static str {
        match self {
            Self::Landscape => "landscape",
            Self::Portrait => "portrait",
        }
    }
}

/// A fixed window: the stage, and the window that holds it plus the bar. Logical px.
#[derive(Clone, Copy, Debug, PartialEq)]
pub struct Window {
    pub stage_w: f64,
    pub stage_h: f64,
    pub width: f64,
    pub height: f64,
}

fn clamp(v: f64, lo: f64, hi: f64) -> f64 {
    v.max(lo).min(hi)
}

/// The window for a work area (logical px) and a title bar, in an orientation.
pub fn window(work_w: f64, work_h: f64, bar: f64, orientation: Orientation) -> Window {
    match orientation {
        Orientation::Landscape => {
            let Constants { k, min, max } = LANDSCAPE;
            let w = clamp(k * work_w, min, max).min((work_h - bar - MARGIN) * RATIO);
            let h = w / RATIO;
            Window {
                stage_w: w,
                stage_h: h,
                width: w,
                height: h + bar,
            }
        }
        Orientation::Portrait => {
            let Constants { k, min, max } = PORTRAIT;
            let h = clamp(k * work_h, min, max).min(work_h - bar - MARGIN);
            let w = h / RATIO;
            Window {
                stage_w: w,
                stage_h: h,
                width: w,
                height: h + bar,
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    /// (work area, bar, landscape stage, portrait stage) — one row of the document's table.
    type Row = ((f64, f64), f64, (f64, f64), (f64, f64));

    /// The contract document's desktop table, verbatim. stage.test.mjs reads the same rows out
    /// of the document and checks them against the JavaScript oracle; this holds the Rust side
    /// to the same numbers.
    const TABLE: &[Row] = &[
        ((1470.0, 850.0), 52.0, (840.0, 630.0), (574.0, 765.0)),
        ((2560.0, 1410.0), 52.0, (1280.0, 960.0), (900.0, 1200.0)),
        ((1366.0, 720.0), 44.0, (840.0, 630.0), (486.0, 648.0)),
        ((1280.0, 672.0), 44.0, (816.0, 612.0), (454.0, 605.0)),
    ];

    #[test]
    fn the_documents_desktop_table_is_what_the_formulas_give() {
        for ((ww, wh), bar, land, port) in TABLE {
            let l = window(*ww, *wh, *bar, Orientation::Landscape);
            let p = window(*ww, *wh, *bar, Orientation::Portrait);
            assert_eq!(
                (l.stage_w.round(), l.stage_h.round()),
                *land,
                "landscape for {ww}×{wh}"
            );
            assert_eq!(
                (p.stage_w.round(), p.stage_h.round()),
                *port,
                "portrait for {ww}×{wh}"
            );
        }
    }

    #[test]
    fn the_window_is_the_stage_plus_the_bar_and_never_exceeds_the_work_area() {
        for ((ww, wh), bar, _, _) in TABLE {
            for o in [Orientation::Landscape, Orientation::Portrait] {
                let w = window(*ww, *wh, *bar, o);
                assert_eq!(w.width, w.stage_w);
                assert_eq!(w.height, w.stage_h + bar);
                assert!(
                    w.width <= *ww && w.height <= *wh,
                    "{o:?} on {ww}×{wh}: {w:?}"
                );
                let ratio = match o {
                    Orientation::Landscape => w.stage_w / w.stage_h,
                    Orientation::Portrait => w.stage_h / w.stage_w,
                };
                assert!((ratio - RATIO).abs() < 1e-9, "stage ratio {ratio}");
            }
        }
    }

    #[test]
    fn a_work_area_too_small_for_the_minimum_gives_the_height_bound_window_not_the_minimum() {
        // The 150%-scaled Windows laptop: 1280×672, bar 44 — below minW, and it must still fit.
        let w = window(1280.0, 672.0, 44.0, Orientation::Landscape);
        assert!(w.stage_w < LANDSCAPE.min);
        assert!(w.height <= 672.0);
    }

    #[test]
    fn orientation_round_trips_through_its_persisted_form_and_defaults_to_landscape() {
        for o in [Orientation::Landscape, Orientation::Portrait] {
            assert_eq!(Orientation::parse(o.as_str()), o);
        }
        assert_eq!(Orientation::parse(""), Orientation::Landscape);
        assert_eq!(Orientation::parse("sideways\n"), Orientation::Landscape);
        assert_eq!(Orientation::parse("portrait\n"), Orientation::Portrait);
    }
}
